File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
// LET TRY BETTER THE BRUTE FORCE APPROACH

// In this approach i am using Extra space o(n+m);
// but the time complexity is O(n*m) which is better than brute force approach



class Solution {
    public void setZeroes(int[][] matrix) {
        
        int n = matrix.length;
        int m = matrix[0].length;
        
        int[] row = new int[n];
        int[] col = new int[m];
        
         for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(matrix[i][j] == 0){
                    row[i] = 1;
                    col[j] = 1;
                }
            }
        }
        
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(row[i] == 1 || col[j] == 1){
                    matrix[i][j] = 0;
                }
            }
        }
        
        
        
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
// let try brute force first 

// TC : O(n^3) 

class Solution {
public:
    
    
    void mark_Row(vector<vector<int>>& matrix, int n,int m,int i){
        
        for(int j = 0; j<m; j++){
            if(matrix[i][j] != 0){
                matrix[i][j] = INT_MIN+1 ;
            }
        }
    }
    
    void mark_col(vector<vector<int>>& matrix,int n, int m, int j){
        
        for(int i = 0; i<n; i++){
            if(matrix[i][j] != 0){
                matrix[i][j] = INT_MIN+1;
            }
        }
    }
    void setZeroes(vector<vector<int>>& matrix) {
        
        int n = matrix.size();
        int m = matrix[0].size();
        
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(matrix[i][j] == 0){
                    mark_Row(matrix,n,m,i);
                    mark_col(matrix,n,m,j);
                }
            }
        }
        
        
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(matrix[i][j] == INT_MIN+1){
                    matrix[i][j] = 0;
                }
            }
        }
    }
};